apache/trafficcontrol
# Add parents to varnish cache

Add the least required changes to run Varnish in a topology with parents by reusing some implementation already made in `go-atscfg` and add Dockerfile for Varnish to be testable in CIAB.

<hr/>

## Which Traffic Control components are affected by this PR?

- Traffic Control Cache Config (`t3c`, formerly ORT)
- CDN in a Box

## What is the best way to verify this PR?

Using CIAB:
- change the dockerfile path to point to varnish dir in either edge or mid.
- change the status to ONLINE in the cache server you choose overriding Traffic Monitor reports. for edge it will be [here](https://github.com/apache/trafficcontrol/blob/master/infrastructure/cdn-in-a-box/traffic_ops/to-access.sh#L217).
- start docker-compose and curl the CDN domain from enroller container.
- check varnish container logs to verify it properly handled the request.
- try again and it should be a hit returning cached object.

Note: it has some issues with mid-02 as the initial data from Traffic Ops does not show any parents or origins. So, ATS `remap.config` and `parent.config` is initially empty unlike edge and mid-01. but, since t3c is running periodically it gets fixed once the data is correct. In Varnish case since t3c-apply is not integrated with Varnish only initial data is used and it is not changed. So, it is not working as mid-02 only mid-01. you can avoid that by adding `sleep` before running t3c-generate to ensure right data exists.

## PR submission checklist
